runtime/parachains/src/disputes.rs

```diff
@@ -516,6 +512,47 @@ impl<BlockNumber: Clone> DisputeStateImporter<BlockNumber> {
 	}
 }
 
+// A filter on a dispute statement set.
+#[derive(PartialEq)]
+#[cfg_attr(test, derive(Debug))]
+enum StatementSetFilter {
+	// Remove the entire dispute statement set.
+	RemoveAll,
+	// Remove the votes with given index from the statement set.
+	RemoveIndices(Vec<usize>),
+}
+
+impl StatementSetFilter {
+	fn filter_statement_set(self, mut statement_set: DisputeStatementSet)
+		-> Option<DisputeStatementSet>
+	{
+		match self {
+			StatementSetFilter::RemoveAll => None,
+			StatementSetFilter::RemoveIndices(mut indices) => {
+				indices.sort();
+				indices.dedup();
+
+				// reverse order ensures correctness
+				for index in indices.into_iter().rev() {
+					statement_set.statements.remove(index);
```

Comment:
as usual, if we don't care about the ordering after this operation, I'd suggest to use `swap_remove` which guarantees overall linear complexity